### PR TITLE
i1675 IT block dep: Use cc to drive cpp and assembler.

### LINKED ISSUE #1675: asm build warnings w/ cross-compiler about IT block deprecation for ARMv8
We should either suppress these or decide we don't need to test these corner cases (probably former is better):

/work/dr/git/src/suite/tests/common/allasm_thumb.asm: Assembler messages:
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:105: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:105: Warning: it blocks of more than one conditional instruction are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:110: Warning: it blocks of more than one conditional instruction are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:116: Warning: it blocks of more than one conditional instruction are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:122: Warning: it blocks of more than one conditional instruction are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:128: Warning: it blocks of more than one conditional instruction are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:148: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:150: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:166: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:166: Warning: it blocks of more than one conditional instruction are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:171: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:188: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:193: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:202: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:208: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:215: Warning: it blocks containing 16-bit Thumb intsructions of the following class are deprecated in ARMv8: Literal loads
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:229: Warning: it blocks containing wide Thumb instructions are deprecated in ARMv8
/work/dr/git/src/suite/tests/common/allasm_thumb.asm:254: Warning: it blocks of more than one conditional instruction are deprecated in ARMv8

### PR DESCRIPTION
**Pull request by fhahn**
This allows us to easily pass in assembler options by setting the
COMPILE_OPTIONS property.

Fixes #1675

### CODE REVIEW

File: make/cpp2asm_support.cmake [hunk 1]
```diff
@@ -303,28 +304,33 @@ if (APPLE)
...
-    # since using cpp now (see above).
-    # FIXME: I tried setting CMAKE_ASM_DEFINE_FLAG to "--defsym " (not clear
-    # how to get =1 in there :should verify it's needed) but <DEFINES>
-    # comes up empty for me.
-    "<CMAKE_ASM_COMPILER> ${ASM_FLAGS} -o <OBJECT> <OBJECT>.s"
-    )
+  if (AARCHXX)
+      set(CMAKE_ASM_COMPILE_OBJECT
+          "${CMAKE_C_COMPILER} -x assembler-with-cpp -c ${ASM_FLAGS} ${CMAKE_CPP_FLAGS} ${rule_flags} ${rule_defs} <SOURCE> -o <OBJECT>")
+    else ()
```
Comment (derekbruening) [inline make/cpp2asm_support.cmake]: indentation doesn't line up

File: make/cpp2asm_support.cmake [hunk 2]
```diff
@@ -303,28 +304,33 @@ if (APPLE)
...
-    # there's no way to transform to --defsym: luckily we don't need them
-    # since using cpp now (see above).
-    # FIXME: I tried setting CMAKE_ASM_DEFINE_FLAG to "--defsym " (not clear
-    # how to get =1 in there :should verify it's needed) but <DEFINES>
-    # comes up empty for me.
-    "<CMAKE_ASM_COMPILER> ${ASM_FLAGS} -o <OBJECT> <OBJECT>.s"
-    )
+  if (AARCHXX)
+      set(CMAKE_ASM_COMPILE_OBJECT
+          "${CMAKE_C_COMPILER} -x assembler-with-cpp -c ${ASM_FLAGS} ${CMAKE_CPP_FLAGS} ${rule_flags} ${rule_defs} <SOURCE> -o <OBJECT>")
```
Comment (derekbruening) [inline make/cpp2asm_support.cmake]: Is this `-x assembler-with-cpp` available in all versions?  Is a flag test needed?

File: make/cpp2asm_support.cmake [hunk 3]
```diff
@@ -303,28 +304,33 @@ if (APPLE)
...
-    # we also aren't passing any <DEFINES> since for one thing
-    # there's no way to transform to --defsym: luckily we don't need them
-    # since using cpp now (see above).
-    # FIXME: I tried setting CMAKE_ASM_DEFINE_FLAG to "--defsym " (not clear
-    # how to get =1 in there :should verify it's needed) but <DEFINES>
-    # comes up empty for me.
-    "<CMAKE_ASM_COMPILER> ${ASM_FLAGS} -o <OBJECT> <OBJECT>.s"
-    )
+  if (AARCHXX)
+      set(CMAKE_ASM_COMPILE_OBJECT
```
Comment (derekbruening) [inline make/cpp2asm_support.cmake]: Indentation looks off

**Comment by derekbruening** [comment]
Is `-x assembler-with-cpp` available with clang?  I'm curious when it was added to gcc.  If it's available in all versions we care about can we simplify and use it everywhere rather than having ARM-only code here.

**Comment by fhahn** [comment]
I'll take a look! There are a still a few issues that need sorting out.

**Comment by fhahn** [comment]
It looks like it has been available in gcc since 3.2 (https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Overall-Options.html) . Not sure since when it is supported in Clang, but clang 3.8 seems to support it.

It looks like using gcc to drive assemler and cpp behaves slightly different. It fails to find the symbols defined in arm.asm, the readelf check fails with ` 3: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND native_plt_call`. The arm.asm.o contains a `native_plt_call` symbol though.

**Comment by fhahn** [comment]
So it looks like it works for AARCHXX now, but unfortunately "just" using `-x assembler-with-cpp` does not work, because it does not allow us to run the newline script after cpp, which is required for some assembler macros.

So there is no huge benefit, but one advantage is that we can easily use Clang's assembler instead of `as`. What do you think?

**Comment by derekbruening** [comment]
So the reason for this change is just to pass a target-specific flag to the assembler?  The flag should show up in `<FLAGS>`.  Did you try just re-enabling that for the assembler command line?  The comment implies it should work on 2.6.4+.

**Comment by fhahn** [comment]
Argh yes that works now, at least after I upgraded the compiler... Not sure what the problem was when I tried earlier.

**Comment by fhahn** [comment]
> So the reason for this change is just to pass a target-specific flag to the assembler? The flag should show up in <FLAGS>. Did you try just re-enabling that for the assembler command line? The comment implies it should work on 2.6.4+.

The problem with that is that `<FLAGS>` is used by CPP already, which does not accept `-wno-warn-deprecated`. And `as` does not accept `-Wa,-wno-warn-deprecated`. By using the compiler to drive both, we can pass in `-Wa` to both

**Comment by derekbruening** [comment]
There is also ` <LANGUAGE_COMPILE_FLAGS>` but I'm not sure what it holds.

A simple option is to just pass this flag to the assembler for all targets if we're not worried about missing some other useful warnings.

Replacing cpp with gcc -E for all platforms as you and #1675 suggest also seems reasonable.

**Comment by fhahn** [comment]
run aarch64 tests

**Comment by fhahn** [comment]
I've now updated the PR to use the compiler to drive preprocessing and assembling when possible. When using clang, we still have to use as directly.

**Comment by fhahn** [comment]
Argh, clang's assembler does not work....

**Comment by derekbruening** [comment]
This seems like a lot of work...one hacky solution is to just hardcode `-wno-warn-deprecated` in the `"<CMAKE_ASM_COMPILER>...` line.

**Comment by derekbruening** [comment]
Did you try `LANGUAGE_COMPILE_FLAGS`?  If it doesn't work we should consider filing a CMake feature request, and maybe submitting a patch implementing it.

**Comment by fhahn** [comment]
> Did you try LANGUAGE_COMPILE_FLAGS? If it doesn't work we should consider filing a CMake feature request, and maybe submitting a patch implementing it.

I couldn't find any documentation for that. I guess we could just add `-wno-warn-deprecated` for ARM. We might want to miss some relevant warnings, but I do not really have any more time to look into the CMAKE specifics here.

**Comment by fhahn** [comment]
Thanks Derek. Switching to drive preprocessing and assembly by the compiler just caused a few headaches, especially with clang, making it not worth it for now. I'll keep the changes somewhere, because it would probably be good to use LLVM's assembler with clang.

**Comment by derekbruening** [comment]
You could keep the changes on a `project-asm-via-cc` shared project branch.
